rust/src/editor.rs

```diff
@@ -179,17 +187,24 @@ impl Editor {
     fn reset_contents(&mut self, new_contents: Rope) {
         self.engine = Engine::new(new_contents);
         self.text = self.engine.get_head();
-        self.dirty = true;
+        self.text_dirty = true;
         self.view.reset_breaks();
         self.set_cursor(0, true);
     }
 
     // render if needed, sending to ui
-    pub fn render(&mut self, tab_ctx: &TabCtx) {
-        if self.dirty {
+    pub fn render(&mut self) {
+        if self.text_dirty || self.view_dirty {
+            let tab_ctx = self.tab_ref.lock().unwrap();
             tab_ctx.update_tab(&self.view.render(&self.text, self.scroll_to));
-            self.dirty = false;
             self.scroll_to = None;
+            if self.text_dirty {
+                for plugin in &self.plugins {
+                    plugin.update();
```

Comment:
Yes, but it won't be the same information as for rendering views. It will be a delta of the string (possibly with the actual string replaced by a byte count when it's too large), an edit type (keyboard insert, delete, paste, undo, redo), the id of the agent doing the editing (user, plugin), the undo group, and the new revision number.

The revision number will be used for CRDT. The RPCs to fetch data give a revision number, and query the snapshot at that revision. In addition, updates are given relative to that revision.

The undo group is mostly so that the plugin can do edits in the same group. Example, if the user types "{", the plugin inserts "}" after, to balance the parentheses, and sets the same undo group. Then when the user undoes the insert "{", both get undone.